text-autospace: Parse ideograph-alpha, ideograph-numeric and normal

vitorroriz opened the pull request:
#### 94f380ede47d26c550c3e09b4c01a719204f50c2
<pre>
text-autospace: Parse ideograph-alpha, ideograph-numeric and normal
<a href="https://bugs.webkit.org/show_bug.cgi?id=277390">https://bugs.webkit.org/show_bug.cgi?id=277390</a>
<a href="https://rdar.apple.com/problem/132860482">rdar://problem/132860482</a>

Reviewed by Tim Nguyen.

We are now parsing ideograph-alpha, ideograph-numeric and normal
as valid values of text-autospace according to specification [1].

We are also creating text-autospace&apos;s own feature flag.

[1]: <a href="https://drafts.csswg.org/css-text-4/#text-autospace-property">https://drafts.csswg.org/css-text-4/#text-autospace-property</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-autospace-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-autospace-valid-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textAutospaceFromStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTextAutospace):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::TextAutospace::TextAutospace):
(WebCore::TextAutospace::isAuto const):
(WebCore::TextAutospace::isNoAutospace const):
(WebCore::TextAutospace::isNormal const):
(WebCore::TextAutospace::hasIdeographAlpha const):
(WebCore::TextAutospace::hasIdeographNumeric const):
(WebCore::TextAutospace::options):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextAutospace):

Canonical link: <a href="https://commits.webkit.org/281719@main">https://commits.webkit.org/281719@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43783a1fb6a7075d00ee719a72f29bf32f483b5a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60766 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64697 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62896 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47801 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11588 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49140 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7855 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62800 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37370 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/52648 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29972 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34057 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9874 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10226 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/53868 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55896 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10172 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66426 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60013 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4710 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10004 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56508 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4731 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52616 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56690 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13551 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3908 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/81768 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35930 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14236 "Found 21 new JSC stress test failures: stress/proxy-get-and-set-recursion-stack-overflow.js.eager-jettison-no-cjit, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-collect-continuously, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-validate-phases, stress/proxy-set.js.bytecode-cache, stress/proxy-set.js.default, stress/proxy-set.js.dfg-eager, stress/proxy-set.js.dfg-eager-no-cjit-validate, stress/proxy-set.js.eager-jettison-no-cjit, stress/proxy-set.js.mini-mode, stress/proxy-set.js.no-cjit-collect-continuously ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37012 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38105 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/platform/text/TextSpacing.h:
```diff
@@ -53,26 +53,37 @@ inline WTF::TextStream& operator<<(WTF::TextStream& ts, const TextSpacingTrim& v
         Auto = 0,
-        NoAutospace // equivalent to None in text-spacing shorthand
+        NoAutospace, // equivalent to None in text-spacing shorthand
+        IdeographAlpa
```
Alpa

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4366,12 +4366,41 @@ RefPtr<CSSValue> consumeTextAutospace(CSSParserTokenRange& range)
+        auto valueID = range.peek().id();
+
+        if ((valueID == CSSValueIdeographAlpha && seenIdeographAlpha) || ( valueID == CSSValueIdeographNumeric && seenIdeographNumeric))
+            return nullptr;
```
This shouldn't be needed, `auto ident = consumeIdent<CSSValueIdeographAlpha, CSSValueIdeographNumeric>(range);` below will return null if an unrecognized ident is parsed.

vitorroriz replied:
This is for checking that `CSSValueIdeographAlpha CSSValueIdeographAlpha` is invalid

nt1m replied:
Ah I misread, thanks.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2028,9 +2028,26 @@ inline TextSpacingTrim BuilderConverter::convertTextSpacingTrim(BuilderState&, c
+            ASSERT_NOT_REACHED();
+            break;
+        }
     }
```
That doesn't appear to allow chaining ideograph-alpha/numeric

nt1m commented on Source/WebCore/platform/text/TextSpacing.h:
```diff
@@ -53,27 +53,44 @@ inline WTF::TextStream& operator<<(WTF::TextStream& ts, const TextSpacingTrim& v
+class TextAutospace {
+public:
+    enum class Type: uint8_t {
+        // NoAutospace = 0, // equivalent to None in text-spacing shorthand
```
Can you drop this line?

nt1m reviewed (commented):
Can you add a link to the spec: https://drafts.csswg.org/css-text-4/#text-autospace-property

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4366,12 +4366,41 @@ RefPtr<CSSValue> consumeTextAutospace(CSSParserTokenRange& range)
+    while (!range.atEnd()) {
+        auto valueID = range.peek().id();
+
+        if ((valueID == CSSValueIdeographAlpha && seenIdeographAlpha) || ( valueID == CSSValueIdeographNumeric && seenIdeographNumeric))
```
```suggestion
        if ((valueID == CSSValueIdeographAlpha && seenIdeographAlpha) || (valueID == CSSValueIdeographNumeric && seenIdeographNumeric))
```

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -387,14 +387,23 @@ static Ref<CSSPrimitiveValue> textSpacingTrimFromStyle(const RenderStyle& style)
+        return CSSPrimitiveValue::create(CSSValueIdeographAlpha);
+    if (textAutospace.hasIdeographNumeric())
+        return CSSPrimitiveValue::create(CSSValueIdeographNumeric);
 
```
It's interesting because `ideograph-alpha ideograph-numeric` serializes as `normal` in `getComputedStyle()`. Can you check with @fantasai if that is correct?

vitorroriz replied:
The reason for that is because `text-autospace: normal` should have the same behavior as `ideograph-alpha ideograph-numeric` but it should serialize as `normal`, so we handle `ideograph-alpha ideograph-numeric` as it would be `normal` and serialize it as `normal` for avoiding having to represent a distinction between these two ways to set `ideograph-alpha ideograph-numeric`. I will ask if that's ok.

nt1m replied:
I know they have the same behavior, but I do wonder if it's OK in the computed style to alias to `normal`.

vitorroriz replied:
@mdubet thinks we should serialize the same way as author has declared. I was trying to avoid having states that are equal, but I guess I'll need a separated representation for normal to return the right computed value here. We also need to update the parsing test later.

nt1m reviewed (approved):
Seems fine as a starter. Can you check with Elika about the pending question?

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2028,11 +2028,31 @@ inline TextSpacingTrim BuilderConverter::convertTextSpacingTrim(BuilderState&, c
+            return { TextAutospace::Type::Auto };
+        if (primtiveValue->valueID() == CSSValueNormal)
+            return { TextAutospace::Type::Normal };
     }
```
typo: primtive

nt1m reviewed (approved):
Please add a link to the spec in the commit message: https://drafts.csswg.org/css-text-4/#propdef-text-autospace

vitorroriz committed:
text-autospace: Parse ideograph-alpha, ideograph-numeric and normal
https://bugs.webkit.org/show_bug.cgi?id=277390
rdar://problem/132860482

Reviewed by Tim Nguyen.

We are now parsing ideograph-alpha, ideograph-numeric and normal
as valid values of text-autospace according to specification [1].

We are also creating text-autospace's own feature flag.

[1]: https://drafts.csswg.org/css-text-4/#text-autospace-property

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-autospace-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-autospace-valid-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textAutospaceFromStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTextAutospace):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::TextAutospace::TextAutospace):
(WebCore::TextAutospace::isAuto const):
(WebCore::TextAutospace::isNoAutospace const):
(WebCore::TextAutospace::isNormal const):
(WebCore::TextAutospace::hasIdeographAlpha const):
(WebCore::TextAutospace::hasIdeographNumeric const):
(WebCore::TextAutospace::options):
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextAutospace):

Canonical link: https://commits.webkit.org/281719@main
